libfwupdplugin/fu-usb-bos-descriptor.c

```diff
@@ -159,6 +183,24 @@ fu_usb_bos_descriptor_parse(FuFirmware *firmware,
 	return TRUE;
 }
 
+static GByteArray *
+fu_usb_bos_descriptor_write(FuFirmware *firmware, GError **error)
+{
+	FuUsbBosDescriptor *self = FU_USB_BOS_DESCRIPTOR(firmware);
+	g_autoptr(FuUsbBosHdr) st = fu_usb_bos_hdr_new();
+	g_autoptr(GBytes) blob = NULL;
+
+	fu_usb_bos_hdr_set_dev_capability_type(st, self->bos_cap.bDevCapabilityType);
+	blob = fu_firmware_get_image_by_id_bytes(firmware, FU_FIRMWARE_ID_PAYLOAD, NULL);
```

Comment:
why not pass error here too?

Reply:
Ahh, blob is optional -- if there's no blob we just fallback to the struct size as the FuUsbBosHdr.length. If you think blob should always be required that's probably sane too.

Reply:
The BOS payload isn't actually required; I think this code is fine.